build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/rewrite/RewritePlugin.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.gradle.internal.rewrite;
+
+import org.gradle.api.Action;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.DependencyResolveDetails;
+import org.gradle.api.artifacts.ModuleVersionSelector;
+import org.gradle.api.file.ProjectLayout;
+import org.gradle.api.plugins.JavaBasePlugin;
+import org.gradle.api.plugins.JavaPluginExtension;
+import org.gradle.api.provider.ProviderFactory;
+
+import javax.inject.Inject;
+
+/**
+ * Adds the RewriteExtension to the current project and registers tasks per-sourceSet.
+ * Only needs to be applied to projects with java sources. No point in applying this to any project that does
+ * not have java sources of its own, such as the root project in a multi-project builds.
+ */
+public class RewritePlugin implements Plugin<Project> {
+
+    public static final String REWRITE_TASKNAME = "rewrite";
+    private ProviderFactory providerFactory;
+    private ProjectLayout projectLayout;
+
+    @Inject
+    public RewritePlugin(ProviderFactory providerFactory, ProjectLayout projectLayout) {
+        this.providerFactory = providerFactory;
+        this.projectLayout = projectLayout;
+    }
+
+    @Override
+    public void apply(Project project) {
+        final RewriteExtension extension = project.getExtensions().create("rewrite", RewriteExtension.class);
+        // Rewrite module dependencies put here will be available to all rewrite tasks
```

Comment:
I guess we could. It depends on what we're using rewrite for as dependencies differ depending on the use case. for this here we could go with defaults I guess

Reply:
Ah, I see. I wonder if we should have a configuration for rewrite itself, and then a separate one for any _additional_ stuff (I assume rule implementations?). That's probably overkill for now though. This can probably stay as-is.